Parent/Parent/ParentAppDelegate.swift

```diff
@@ -118,27 +92,132 @@ class ParentAppDelegate: UIResponder, UIApplicationDelegate {
     @objc func routeToRemindable(from response: UNNotificationResponse) {
         let userInfo = response.notification.request.content.userInfo
         if let urlString = userInfo[RemindableActionURLKey] as? String, let url = URL(string: urlString) {
-            Router.sharedInstance.route(visibleController, toURL: url, modal: true)
+            Router.sharedInstance.route(topMostViewController()!, toURL: url, modal: true)
+        }
+    }
+
+    func setup(session: KeychainEntry) {
+        environment.userDidLogin(session: session)
+        CoreWebView.keepCookieAlive(for: environment)
+        // UX requires that students are given color schemes in a specific order.
+        // The method call below ensures that we always start with the first color scheme.
+        ColorCoordinator.clearColorSchemeDictionary()
+        if Locale.current.regionCode != "CA" {
+            let crashlyticsUserId = "\(session.userID)@\(session.baseURL.host ?? session.baseURL.absoluteString)"
+            Crashlytics.sharedInstance().setUserIdentifier(crashlyticsUserId)
+        }
+        // Legacy CanvasKeymaster support
+        let legacyClient = CKIClient(baseURL: session.baseURL, token: session.accessToken)!
+        legacyClient.actAsUserID = session.actAsUserID
+        legacyClient.originalIDOfMasqueradingUser = session.originalUserID
+        legacyClient.originalBaseURL = session.originalBaseURL
+        legacyClient.fetchCurrentUser().subscribeNext { user in
+            legacyClient.setValue(user, forKey: "currentUser")
+            CanvasKeymaster.the().setup(with: legacyClient)
+            self.legacySession = legacyClient.authSession
+            Router.sharedInstance.session = legacyClient.authSession
+            NotificationCenter.default.post(name: .loggedIn, object: self, userInfo: [LoggedInNotificationContentsSession: legacyClient.authSession])
+            self.showRootView()
+        }
+    }
+
+    func showRootView() {
+        guard let legacySession = legacySession else { return }
+        do {
+            let refresher = try Student.observedStudentsRefresher(legacySession)
+            refresher.refreshingCompleted.observeValues { [weak self] _ in
```

Comment:
Interesting that this works since `refresher` will go out of scope. I guess `observeValues` will retain the block until the signal complete, maybe.

Reply:
Why are we needing to refresh students before showing the dashboard?

Reply:
Oh, I see further down that this is what we've been doing. Super weird and gross but beyond the scope of this.

Reply:
yeah, 🤷‍♂. I'm trying not to redo _everything_ I touch. Just *most* things. 😉